src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,465 @@
+}
+
+bool QubitListInSharedArray::IsEmpty() const
+{
+    return firstElement == NoneMarker;
+}
+
+void QubitListInSharedArray::AddQubit(QubitIdType id, bool addToFront, QubitIdType* sharedQubitStatusArray)
+{
+    FailIf(id == NoneMarker, "Incorrect qubit id, cannot add it to the list.");
+
+    // If the list is empty, we initialize it with the new element.
+    if (IsEmpty())
+    {
+        firstElement = id;
+        lastElement = id;
+        sharedQubitStatusArray[id] = NoneMarker; // List with a single elemenet in the chain.
+        return;
+    }
+
+    if (addToFront)
+    {
+        sharedQubitStatusArray[id] = firstElement; // The new element will point to the former first element.
+        firstElement = id; // The new element is now the first in the chain.
+    } else
+    {
+        sharedQubitStatusArray[lastElement] = id; // The last element will point to the new element.
+        sharedQubitStatusArray[id] = NoneMarker; // The new element will end the chain.
+        lastElement = id; // The new element will be the last element in the chain.
+    }
+}
+
+// TODO: Set status to the taken qubit here. Then counting is reasonable here, but not possible?
+// TODO: Rename 'RemoveQubitFromFront'?
+int QubitListInSharedArray::TakeQubitFromFront(QubitIdType* sharedQubitStatusArray)
+{
+    // First element will be returned. It is 'NoneMarker' if the list is empty.
+    int result = firstElement;
+
+    // Advance list start to the next element if list is not empty.
+    if (!IsEmpty())
+    {
+        firstElement = sharedQubitStatusArray[firstElement]; // The second element will be the first.
+    }
+
+    // Drop pointer to the last element if list becomes empty.
+    if (IsEmpty())
+    {
+        lastElement = NoneMarker;
+    }
+
+    return result;
+}
+
+void QubitListInSharedArray::MoveAllQubitsFrom(QubitListInSharedArray& source, QubitIdType* sharedQubitStatusArray)
+{
+    // No need to do anthing if source is empty.
+    if (source.IsEmpty())
+    {
+        return;
+    }
+
+    if (this->IsEmpty())
+    {
+        // If this list is empty, we'll just set it to the source list.
+        lastElement = source.lastElement;
+    } else
+    {
+        // Attach source at the beginning of the list if both lists aren't empty.
+        sharedQubitStatusArray[source.lastElement] = firstElement; // The last element of the source chain will point to the first element of this chain.
+    }
+    firstElement = source.firstElement; // The first element from the source chain will be the first element of this chain.
+
+    // Remove all elements from source.
+    source.firstElement = NoneMarker;
+    source.lastElement = NoneMarker;
+}
+
+
+//
+// RestrictedReuseArea
+//
+
+RestrictedReuseArea::RestrictedReuseArea(QubitListInSharedArray freeQubits)
+{
+    //FreeQubitsReuseProhibited = QubitListInSharedArray(); // This is initialized by default.
+    FreeQubitsReuseAllowed = freeQubits; // Default shallow copying.
+}
```

Comment:
(Up to you)
Consider
* Making the parameter a reference to `const`
* And using the member initialization list in all the constructors.

```c++
RestrictedReuseArea::RestrictedReuseArea(const QubitListInSharedArray& freeQubits) // Param is ref to const
    : FreeQubitsReuseAllowed(freeQubits)  // Member initialization list.
{    
}
```
See [Item 4: Make sure that objects are initialized before they're used](http://wavelino.coffeecup.com/pdf/Effective%20Cshapshap.pdf).

Reply:
No need to make it reference to const. This is a small class, think of it as an int or long long. No one would do this for ints, as they are intended to be copied.
As for member initialization - I will think about it. I know the reasons, but I don't like the style.

Reply:
Added non-static data member initialization and initialization lists where appropriate.